### PR TITLE
add writeup of how model samples a new name

### PR DESCRIPTION
**Pull request by nathandalal**
This changeset includes a new section in the GUI `streamlit` app that displays how the model does inference.
This will be useful for beginning learners who need a grasp on probability and sampling, and need a visual to help explain this concept.

I added this because I'm excited about this project, and additionally I believe that the GUI needed an emphasis on inference and how it works, alongside the training and hyperparameter optimization process.

Here is an accompanying video showing how the process works.

https://github.com/user-attachments/assets/cd10b050-23f0-45d2-ac43-c4a47e4531d1

I have added the content under the section "How the n-gram model generates names". Cheers!

### CODE REVIEW

**Comment by lakshyaag** [comment]
Nice! I added something similar here: [ash-01xor#1](https://github.com/ash-01xor/ngram/pull/1)

**Comment by nathandalal** [comment]
Sweet! This is awesome. Thanks @lakshyaag. We'll all need a bit more project planning / guidance from @karpathy on how to set up some more foundational things here so we can all quickly contribute in a structured way. This is just the beginning :)

File: gui.py [hunk 1]
```diff
@@ -59,21 +60,22 @@
...
-    vocab_size = len(uchars)
-    char_to_token = {c: i for i, c in enumerate(uchars)}
-    token_to_char = {i: c for i, c in enumerate(uchars)}
-    # designate \n as the delimiting <|endoftext|> token
-    EOT_TOKEN = char_to_token["\n"]
-    # pre-tokenize all the splits one time up here
-    test_tokens = [char_to_token[c] for c in open(test_tokens_path, "r").read()]
-    val_tokens = [char_to_token[c] for c in open(val_tokens_path, "r").read()]
-    train_tokens = [char_to_token[c] for c in open(train_tokens_path, "r").read()]
+# Foundationally useful code for all visualizations. Doing tokenization.
```
Comment (jackos) [inline gui.py:63]: Yes good call thanks

File: gui.py [hunk 2]
```diff
@@ -203,3 +205,88 @@
     This visualization helps to understand how the model has learned the statistical 
     patterns of the language from the training data.
     """)
+
+    st.write("## How the N-Gram model generates names")
```
Comment (jackos) [inline gui.py:209]: This is great cheers!

File: ngram.py [hunk 3]
```diff
@@ -128,6 +128,7 @@ def eval_split(model, tokens):
 
 # -----------------------------------------------------------------------------
 
+
```
Comment (jackos) [inline ngram.py:131]: He uses 1 line spacing between everything in this file.

File: ngram.py [hunk 4]
```diff
@@ -156,7 +157,7 @@ def main():
         train_loss = eval_split(model, train_tokens)
         val_loss = eval_split(model, val_tokens)
         print("seq_len %d | smoothing %.2f | train_loss %.4f | val_loss %.4f"
-            % (seq_len, smoothing, train_loss, val_loss))
+              % (seq_len, smoothing, train_loss, val_loss))
```
Comment (jackos) [inline ngram.py:160]: This changes it to 5 spaces, can it be changed back to 4?

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1,200 @@
+aiohttp==3.8.4
```
Comment (jackos) [inline requirements.txt:1]: If you look at his other projects he [doesn't commit a frozen `requirements.txt`](https://github.com/dorjeduck/llm.mojo/blob/main/requirements.txt), and likes to keep things simple. He also doesn't lock the versions. Just need to have these here:

```
matplotlib
streamlit
seaborn
```

Other things like numpy are a dependency of streamlit

**Review by jackos — CHANGES_REQUESTED** [review]
Awesome looks great thanks @nathandalal! Have a couple of changes based on his other projects.

File: requirements.txt (hunk 5 above)
Comment (nathandalal) [reply requirements.txt:1]: Got it yeah; it must be helpful for students to quite literally "spell out" requirements as they go.

**Pull request merged**